serverless/client/witness/consensus.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package witness provides client support for using witnesses.
+package witness
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian-examples/serverless/client"
+	"golang.org/x/mod/sumdb/note"
+)
+
+// CheckpointNConsensus returns a ConsensusCheckpoint function which selects the newest checkpoint available from
+// the available distributors which has signatures from at least N of the provided witnesses.
+func CheckpointNConsensus(logID string, distributors []client.Fetcher, witnesses []note.Verifier, N int) client.ConsensusCheckpoint {
+	// TODO(al): This implementation is pretty basic, and could be made better.
+	// In particular, there are cases where it will fail to build consensus
+	// where a more thorough algorithm would succeed.
+	// e.g. currently, it will simply fetch  checkpoint.N from each of the
+	// distributors and fail if either:
+	//  - no distributor has a checkpoint.N file
+	//  - no distributor has a checkpoint.N file signed by N of the known witnesses.
+	//
+	// It's good enough for now, though.
+	return func(ctx context.Context, logSigV note.Verifier, origin string) (*log.Checkpoint, []byte, error) {
```

Comment:
```suggestion
        if N > len(witnesses) { panic("plonker!"); }
	return func(ctx context.Context, logSigV note.Verifier, origin string) (*log.Checkpoint, []byte, error) {
```

Reply:
panicking seems a bit rude in a lib, but I've updated to return `error` too.